### ISSUE TITLE
help request: Failed to parse domain

### ISSUE THREAD

**Issue by Kozatoo** [apache/apisix#10505]
### Description

Context: I'm trying to build an application with an Auth Service which should only validate the tokens, and admin service which sould be protected. 

I am running APIsix from docker container, but I keep getting errors about failing to parse the domain.
When runing echo $variable_name inside the container, it echoes perfectly. But I still get these errors inside the log file.

The calls to the Auth service to check for the API are actually working, and the call to Admin service as well. But I still get these errors in the logs

On another hand, when effectuating these API calls from a front-end application, I get a CORS error, and I noted that the authorization header is not present inside the request sent with forward-auth plugin (even though I've added it inside the request_headers) 

Here are the logs

```

2023/11/16 11:08:54 [error] 55#55: *160063 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}}, error: failed to query the DNSserver: dns server error: 3 name error, client: 172.21.0.1, server: _, request: "OPTIONS /auth/greet HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:4200/"
2023/11/16 11:08:54 [error] 55#55: *160063 [lua] upstream.lua:79: parse_domain_for_nodes(): dns resolver domain: ${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}} error: failed to query the DNS server: dns server error: 3 name error, client: 172.21.0.1, server: _, request: "OPTIONS /auth/greet HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:4200/"
2023/11/16 11:08:54 [error] 55#55: *160063 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}}, error: failed to query the DNSserver: dns server error: 3 name error, client: 172.21.0.1, server: _, request: "GET /auth/greet HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:4200/"
2023/11/16 11:08:54 [error] 55#55: *160063 [lua] upstream.lua:79: parse_domain_for_nodes(): dns resolver domain: ${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}} error: failed to query the DNS server: dns server error: 3 name error, client: 172.21.0.1, server: _, request: "GET /auth/greet HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:4200/"
2023/11/16 11:08:55 [warn] 55#55: *160063 [lua] plugin.lua:1107: run_plugin(): forward-auth exits with http status code 401, client: 172.21.0.1, server: _, request: "OPTIONS /admin/users HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:4200/"
```

apisix.yaml
```

routes:
  - uri: /auth/*
    plugins:
      proxy-rewrite:
        regex_uri: ["^/auth/(.*)","/$1"]

    upstream:
      nodes:
        "${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}}": 1
      type: roundrobin
    
  - uri: /admin/*
    plugins:  
      proxy-rewrite:
        regex_uri: ["^/admin/(.*)","/$1"]
      forward-auth:
        uri: http://${{AUTH_SERVICE_IP:=}}:${{AUTH_SERVICE_PORT:=3000}}/validate
        ssl_verify: false
        request_headers: ["Authorization",authorization]
    upstream:
      nodes:
        "${{ADMIN_SERVICE_IP:=192.168.1.76}}:${{ADMIN_SERVICE_PORT:=3001}}": 1
    
#END

```

### Environment

APISIX: 3.6 dockerized
etcd:3.4.15 dockerized

**Comment by markokocic** [comment]
Looks like your issue is that the standalone configuration file `apisix.yml` doesn't support usage of environment variables. It assumes your host name is your actual variable name.

Maybe this would be a nice feature for Apisix, to bring standalone config file up to date with other config files.

**Comment by Kozatoo** [comment]
This isn't mentioned in the docs, I guess it should be updated

**Comment by kayx23** [comment]
I remember a PR from a long time ago that supported env var in `apisix.yaml`.

Found the test here: https://github.com/apache/apisix/blob/master/t/cli/test_standalone.sh#L39-L51

**Comment by monkeyDluffy6017** [comment]
@Kozatoo have you solved your problem?

**Comment by shreemaan-abhishek** [comment]
closing in favor of https://github.com/apache/apisix/issues/10530#issuecomment-1876222842

**Comment by ilya-tishonok** [comment]
@shreemaan-abhishek, it seems like issue is not connected with ```apisix.yaml``` config changing (as described in #10530), we are facing same issue at initial startup (without chaning apisix.yaml) in docker environment.
We have the following apisix.yaml configuration:

```yaml
upstreams:
  -
    id: some_service_upstream
    nodes:
      "${{SOME_SERVICE_UPSTREAM_ADDRESS}}": 1
    type: roundrobin
```

and docker-compose file:

```yaml
version: "3"

services:
  apisix:
    build:
      context: ../../gateway
      dockerfile: ../docker/Dockerfile.public-api-gateway
    container_name: apisix_api_gateway
    restart: always
    environment:
      # Upstream Addresses Envs Start
      LOS_SERVICE_UPSTREAM_ADDRESS: ${LOS_SERVICE_UPSTREAM_ADDRESS}
    ports:
      - "9180:9180/tcp"
      - "9080:9080/tcp"
      - "9091:9091/tcp"
      - "9444:9443/tcp"
      - "9092:9092/tcp"
    networks:
      apisix:

networks:
  apisix:
    driver: bridge
```
    
 while doing requests to LOS service we are seeing this in console:
   
![image](https://github.com/apache/apisix/assets/67690609/462a0f81-5a6d-4449-9546-13e7cdb88f92)

but request is hitting upstream service.

**Comment by ZappaBoy** [comment]
> @shreemaan-abhishek, it seems like issue is not connected with `apisix.yaml` config changing (as described in #10530), we are facing same issue at initial startup (without chaning apisix.yaml) in docker environment. We have the following apisix.yaml configuration:
> 
> ```yaml
> upstreams:
>   -
>     id: some_service_upstream
>     nodes:
>       "${{SOME_SERVICE_UPSTREAM_ADDRESS}}": 1
>     type: roundrobin
> ```
> 
> and docker-compose file:
> 
> ```yaml
> version: "3"
> 
> services:
>   apisix:
>     build:
>       context: ../../gateway
>       dockerfile: ../docker/Dockerfile.public-api-gateway
>     container_name: apisix_api_gateway
>     restart: always
>     environment:
>       # Upstream Addresses Envs Start
>       LOS_SERVICE_UPSTREAM_ADDRESS: ${LOS_SERVICE_UPSTREAM_ADDRESS}
>     ports:
>       - "9180:9180/tcp"
>       - "9080:9080/tcp"
>       - "9091:9091/tcp"
>       - "9444:9443/tcp"
>       - "9092:9092/tcp"
>     networks:
>       apisix:
> 
> networks:
>   apisix:
>     driver: bridge
> ```
> 
> while doing requests to LOS service we are seeing this in console:
> 
> ![image](https://private-user-images.githubusercontent.com/67690609/335691771-462a0f81-5a6d-4449-9546-13e7cdb88f92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc0MDc5NTUsIm5iZiI6MTcxNzQwNzY1NSwicGF0aCI6Ii82NzY5MDYwOS8zMzU2OTE3NzEtNDYyYTBmODEtNWE2ZC00NDQ5LTk1NDYtMTNlN2NkYjg4ZjkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjAzVDA5NDA1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM4ZWVkYTBkYjgzYTBjZTM2MzZmZmQ1NGJlNzdhZWIyYzkzMzNiZTA2MzhjZjc4YjZlMmZkMTNlZDViNTVkNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.S_vBxv5jmCiTI2a0V1tNaNR8jgPkZ-KzDP0n87AGdaM)
> 
> but request is hitting upstream service.

@ilya-tishonok The same issue here. I also tried with different `apisix` versions, specifying `dns_resolver` property and changing `apisix.yaml` settings but the issue still appears. The services return the responses correctly and the responses are also cached by the `proxy-cache` plugin but the `parse_domain_for_nodes()` function always gives the same error.

**Comment by wingyiu** [comment]
```
[lua] resolver.lua:80: parse_domain(): failed to parse domain: prod-k8s-web-gpt.k8s-web.svc.cluster.local, error: failed to query the DNS server: dns server error: 3 name error
```

**Comment by giangnh13579** [comment]
> @shreemaan-abhishek, it seems like issue is not connected with `apisix.yaml` config changing (as described in #10530), we are facing same issue at initial startup (without chaning apisix.yaml) in docker environment. We have the following apisix.yaml configuration:
> 
> ```yaml
> upstreams:
>   -
>     id: some_service_upstream
>     nodes:
>       "${{SOME_SERVICE_UPSTREAM_ADDRESS}}": 1
>     type: roundrobin
> ```
> 
> and docker-compose file:
> 
> ```yaml
> version: "3"
> 
> services:
>   apisix:
>     build:
>       context: ../../gateway
>       dockerfile: ../docker/Dockerfile.public-api-gateway
>     container_name: apisix_api_gateway
>     restart: always
>     environment:
>       # Upstream Addresses Envs Start
>       LOS_SERVICE_UPSTREAM_ADDRESS: ${LOS_SERVICE_UPSTREAM_ADDRESS}
>     ports:
>       - "9180:9180/tcp"
>       - "9080:9080/tcp"
>       - "9091:9091/tcp"
>       - "9444:9443/tcp"
>       - "9092:9092/tcp"
>     networks:
>       apisix:
> 
> networks:
>   apisix:
>     driver: bridge
> ```
> 
> while doing requests to LOS service we are seeing this in console:
> 
> ![image](https://private-user-images.githubusercontent.com/67690609/335691771-462a0f81-5a6d-4449-9546-13e7cdb88f92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1NTA4MDUsIm5iZiI6MTczNTU1MDUwNSwicGF0aCI6Ii82NzY5MDYwOS8zMzU2OTE3NzEtNDYyYTBmODEtNWE2ZC00NDQ5LTk1NDYtMTNlN2NkYjg4ZjkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDA5MjE0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5OWU3N2RlOThhNmQyYjI5MjcyZWNjYTllYTlkN2YyZmNhNTBkNjc0MTMyYTliYTVkNjY5NjIxMTgwZDllNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K5LqYYxzOA2xjqn_LdqOYPBHeBJ9IRoFxpO1BzFqwiE)
> 
> but request is hitting upstream service.

I'm using APISIX `3.11` in the docker environment and got the same error. Does this issue have any solution?

**Comment by berkaycirak** [comment]
any update

**Comment by perry2of5** [comment]
I'm having this issue using the standard docker container. I finally just put in a script to run before apisix starts to expand the variables.

```bash
#!/bin/bash

# for whatever reason, apisix isn't expanding the environment variables in apisix.yaml
# Instead, we'll just do it before apisix starts. This is a hack, but until we can
# figure out why apisix isn't doing it, this will have to do.

cd ${APISIX_CONF_DIR}

env | grep '^KNOWN_PREFIX_' | grep _SVC | while IFS='=' read -r name value ; do
  echo expanding $name
  # using sed to replace variable in format ${{VAR_NAME}} in apisix.yaml in this directory
  sed -i "s|\${{$name}}|$value|g" apisix.yaml
done
```

I have checked and when the apisix docker container starts it is putting the correct env entries in nginx.conf 

It seems like there is an issue around expanding environment variables in apisix.yaml that hasn't been address. Is the right thing to open a new ticket with an example project? What is the useful way to move this forward? I'm feeling sure commenting on a closed ticket might not be the most useful. However, I am providing some evidence the environment variables are making it into the docker container and that the nginx.conf file has the right env entries so maybe that is enough? Let me know! Thanks for the great tool.

**Comment by ZappaBoy** [comment]
@perry2of5 Nice, thank you. 

Honestly, I remember noticing something similar, but I don't remember how I bypassed this error one year ago. I can't understand why something as basic and useful as this has not been addressed in one year by `apisix`. 

By the way, I think that you should open a Pull Request adding the `nginx.conf` to the `docker` folder ([here](https://github.com/apache/apisix/tree/master/docker/compose)). This should be helpful for those who have this issue.

**Comment by perry2of5** [comment]
Here is a minimal reproduction: https://github.com/perry2of5/apisix-docker


`git clone git@github.com:perry2of5/apisix-docker.git`
`cd apisix-docker`
`docker build -t apisix-demo .`
`docker run -p 8080:8080 apisix-demo`

## route with variable not in hostname works without errors
In a separate window, works without errors:

```
% curl http://localhost:8080/api/supabase-config/v1 
{"supabase_url":"https://somehashvalue.supabase.co", "supabase_publishable_key":"blah-blah-blah"}
```
accompanying apisix log:
```
192.168.65.1 - - [16/Sep/2025:05:05:49 +0000] localhost:8080 "GET /api/supabase-config/v1 HTTP/1.1" 200 108 0.000 "-" "curl/8.7.1" - - - "http://localhost:8080"
```

## hostname shows errors
```
% curl http://localhost:8080/api/somename/v1/always/there/look/ma
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Host": "httpbin.org", 
    "User-Agent": "curl/8.7.1", 
    "X-Amzn-Trace-Id": "Root=1-68c8f042-5a6cd9265f935e2b422a46eb", 
    "X-Forwarded-Host": "localhost"
  }, 
  "json": null, 
  "method": "GET", 
  "origin": "192.168.65.1, 108.204.120.162", 
  "url": "http://localhost/anything/always/there/look/ma"
}
```
And the apisix logs with the misleading errors:
```
2025/09/16 05:06:09 [error] 52#52: *2855 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{HTTPBIN_SERVER}}, error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
2025/09/16 05:06:09 [error] 52#52: *2855 [lua] upstream.lua:97: parse_domain_for_nodes(): dns resolver domain: ${{HTTPBIN_SERVER}} error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
192.168.65.1 - - [16/Sep/2025:05:06:10 +0000] localhost:8080 "GET /api/somename/v1/always/there/look/ma HTTP/1.1" 200 414 1.038 "-" "curl/8.7.1" 54.227.118.219:80 200 0.930 "http://httpbin.org/anything/always/there/look/ma"
```

**Comment by arulrajnet** [comment]
I am using apisix docker verison `3.13.0-debian`. Got the following error, but the request is proxied to the upstream successfully. 

```
2025/09/19 09:31:18 [error] 33#33: *82474 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{API_SERVICE_HOST}}, error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/v1/users/me HTTP/2.0", host: "localhost"
2025/09/19 09:31:18 [error] 33#33: *82474 [lua] upstream.lua:97: parse_domain_for_nodes(): dns resolver domain: ${{API_SERVICE_HOST}} error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/v1/users/me HTTP/2.0", host: "localhost"
``` 

The apisix.yaml has the following nodes

```
services:
  - id: 2
    name: whoami-service
    upstream:
      type: roundrobin
      nodes:
        "${{API_SERVICE_HOST}}:${{API_SERVICE_PORT}}": 1
```

Is there any workaround to avoid this error log? My logs are having full of this log.

**Comment by perry2of5** [comment]
As mentioned above, I'm expanding the variables before starting apisix. I added "tini" to ensure signals make it to APISIX and then I have startup script "/usr/local/bin/startup.bash"

```bash
#!/bin/bash
pushd ${APISIX_CONF_DIR}

env | grep '^KNOWN_PREFIX_' | grep _SVC | while IFS='=' read -r name value ; do
  echo expanding $name
  # using sed to replace variable in format ${{VAR_NAME}} in apisix.yaml in this directory
  sed -i "s|\${{$name}}|$value|g" apisix.yaml
done


popd
/docker-entrypoint.sh "$@" &
APISIX_PID=$!
echo "Started APISIX with PID $APISIX_PID"

trap "kill $APISIX_PID; exit" SIGTERM SIGINT

wait $NPM_PID
wait $APISIX_PID
```


Then in my Dockerfile:

```docker
ENTRYPOINT ["/usr/bin/tini", "--", "/usr/local/bin/startup.bash"]

CMD ["docker-start"]
STOPSIGNAL SIGQUIT
```

Feedback on ways to improve are welcome :)

**Comment by Baoyuantop** [comment]
> Here is a minimal reproduction: https://github.com/perry2of5/apisix-docker
> 
> `git clone git@github.com:perry2of5/apisix-docker.git` `cd apisix-docker` `docker build -t apisix-demo .` `docker run -p 8080:8080 apisix-demo`
> 
> ## route with variable not in hostname works without errors
> In a separate window, works without errors:
> 
> ```
> % curl http://localhost:8080/api/supabase-config/v1 
> {"supabase_url":"https://somehashvalue.supabase.co", "supabase_publishable_key":"blah-blah-blah"}
> ```
> 
> accompanying apisix log:
> 
> ```
> 192.168.65.1 - - [16/Sep/2025:05:05:49 +0000] localhost:8080 "GET /api/supabase-config/v1 HTTP/1.1" 200 108 0.000 "-" "curl/8.7.1" - - - "http://localhost:8080"
> ```
> 
> ## hostname shows errors
> ```
> % curl http://localhost:8080/api/somename/v1/always/there/look/ma
> {
>   "args": {}, 
>   "data": "", 
>   "files": {}, 
>   "form": {}, 
>   "headers": {
>     "Accept": "*/*", 
>     "Host": "httpbin.org", 
>     "User-Agent": "curl/8.7.1", 
>     "X-Amzn-Trace-Id": "Root=1-68c8f042-5a6cd9265f935e2b422a46eb", 
>     "X-Forwarded-Host": "localhost"
>   }, 
>   "json": null, 
>   "method": "GET", 
>   "origin": "192.168.65.1, 108.204.120.162", 
>   "url": "http://localhost/anything/always/there/look/ma"
> }
> ```
> 
> And the apisix logs with the misleading errors:
> 
> ```
> 2025/09/16 05:06:09 [error] 52#52: *2855 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{HTTPBIN_SERVER}}, error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
> 2025/09/16 05:06:09 [error] 52#52: *2855 [lua] upstream.lua:97: parse_domain_for_nodes(): dns resolver domain: ${{HTTPBIN_SERVER}} error: failed to query the DNS server: dns server error: 3 name error, client: 192.168.65.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
> 192.168.65.1 - - [16/Sep/2025:05:06:10 +0000] localhost:8080 "GET /api/somename/v1/always/there/look/ma HTTP/1.1" 200 414 1.038 "-" "curl/8.7.1" 54.227.118.219:80 200 0.930 "http://httpbin.org/anything/always/there/look/ma"
> ```

It's a bit strange that I didn't see the expected error log after following the steps here.

**Comment by perry2of5** [comment]
I wonder if it is architecture related. I'm running the latest docker desktop on an M4 series Mac and the container is running arm64 according to docker image inspect.

Docker Desktop v4.46.0
macOS Sequoia 15.6.1

**Comment by perry2of5** [comment]
I was able to reproduce on a windows 10 pro computer running docker desktop 4.32.0.

```
172.17.0.1 - - [22/Sep/2025:18:23:59 +0000] localhost:8080 "GET /api/supabase-config/v1 HTTP/1.1" 200 108 0.000 "-" "curl/8.13.0" - - - "http://localhost:8080"
2025/09/22 18:24:16 [error] 54#54: *2524 [lua] resolver.lua:80: parse_domain(): failed to parse domain: ${{HTTPBIN_SERVER}}, error: failed to query the DNS server: dns server error: 3 name error, client: 172.17.0.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
2025/09/22 18:24:16 [error] 54#54: *2524 [lua] upstream.lua:97: parse_domain_for_nodes(): dns resolver domain: ${{HTTPBIN_SERVER}} error: failed to query the DNS server: dns server error: 3 name error, client: 172.17.0.1, server: _, request: "GET /api/somename/v1/always/there/look/ma HTTP/1.1", host: "localhost:8080"
172.17.0.1 - - [22/Sep/2025:18:24:18 +0000] localhost:8080 "GET /api/somename/v1/always/there/look/ma HTTP/1.1" 200 413 1.175 "-" "curl/8.13.0" 44.195.242.49:80 200 1.138 "http://httpbin.org/anything/always/there/look/ma"
```

**Comment by perry2of5** [comment]
Did you run the second curl command?

```bash
 curl http://localhost:8080/api/somename/v1/always/there/look/ma
```

**Comment by Baoyuantop** [comment]
I have reproduced this issue on an x86 Ubuntu machine, and I will continue to investigate possible causes of the error.

**Comment by nic-6443** [comment]
This got fixed in two halves. #13078 (shipped in 3.17.0) moves env substitution before YAML parsing, which fixes the parse_domain error reported here for standalone apisix.yaml. The leftover path — the old `${{VAR}}` key never being removed because of a `conf.key = nil` typo (should have been `conf[key] = nil`) — was fixed by #12885, merged this week and shipping with the next release later this month.

Upgrading to 3.17.0 already makes the reported symptom go away. Closing.

**Comment by perry2of5** [comment]
Thank you.

**Comment by tbailly** [comment]
Just came across this issue, thanks for the efforts in resolving this!
